extensions/mbw/xml.js

```diff
@@ -0,0 +1,598 @@
+            arguments: {
+              QUERY: {
+                type: Scratch.ArgumentType.STRING,
+                defaultValue: ".foo",
+              },
+              XML: {
+                type: Scratch.ArgumentType.STRING,
+                defaultValue: '<a><b /><c class="foo" /></a>',
+              },
+            },
+          },
+          {
+            opcode: "querySelector",
+            blockType: Scratch.BlockType.REPORTER,
+            text: Scratch.translate("query [QUERY] on [XML]"),
+            arguments: {
+              QUERY: {
+                type: Scratch.ArgumentType.STRING,
+                defaultValue: ".foo",
+              },
+              XML: {
+                type: Scratch.ArgumentType.STRING,
+                defaultValue: '<a><b /><c class="foo" /></a>',
+              },
+            },
+          },
+          {
+            opcode: "querySelectorAll",
+            blockType: Scratch.BlockType.REPORTER,
+            text: Scratch.translate("query all [QUERY] on [XML]"),
+            arguments: {
+              QUERY: {
+                type: Scratch.ArgumentType.STRING,
+                defaultValue: ".foo",
+              },
+              XML: {
+                type: Scratch.ArgumentType.STRING,
+                defaultValue: '<a><b class="foo"/><c class="foo" /></a>',
+              },
+            },
+          },
+        ],
+      };
+    }
+
+    /**
+     * @param {object} args
+     * @param {unknown} args.MAYBE_XML
+     */
+    isValid({ MAYBE_XML }) {
+      const { xml } = this.stringToXml(Scratch.Cast.toString(MAYBE_XML));
+      return xml !== null;
+    }
+
+    /**
+     * @param {object} args
+     * @param {unknown} args.MAYBE_XML
+     */
+    errorMessage({ MAYBE_XML }) {
+      const { xml, error } = this.stringToXml(Scratch.Cast.toString(MAYBE_XML));
+      return xml === null ? error : "";
+    }
+
+    /**
+     * @param {object} args
+     * @param {unknown} args.XML
+     */
+    tagName({ XML }) {
+      const { xml } = this.stringToXml(Scratch.Cast.toString(XML));
+      if (xml === null) {
+        return "";
+      }
+      return xml.tagName;
+    }
+
+    /**
+     * @param {object} args
+     * @param {unknown} args.XML
+     */
+    textContent({ XML }) {
+      const { xml } = this.stringToXml(Scratch.Cast.toString(XML));
+      if (xml === null) {
+        return "";
+      }
+      return xml.textContent;
+    }
+
+    /**
+     * @param {object} args
+     * @param {unknown} args.XML
+     */
+    attributes({ XML }) {
+      const { xml } = this.stringToXml(Scratch.Cast.toString(XML));
+      if (xml === null) {
+        return "";
+      }
+      return JSON.stringify([...xml.attributes].map((attr) => attr.name));
+    }
+
+    /**
+     * @param {object} args
+     * @param {unknown} args.XML
+     * @param {unknown} args.ATTR
+     */
+    hasAttribute({ XML, ATTR }) {
+      return this.getAttribute({ XML, ATTR }) === "";
```

Comment:
Shouldn't this be the opposite? It's checking to see if an attribute exists inside an existing XML script, yet it returns true if getAttribute is null (meaning it doesn't have the attribute).

Reply:
Thanks, you're right